Revamp local state tutorial chapter for AC3 (#1050)
apollographql/apollo: docs/source/tutorial/local-state.mdx

import Disclaimer from '../../shared/disclaimer.mdx';
import {
  ExpansionPanel,
} from 'gatsby-theme-apollo-docs/src/components/expansion-panel';

Time to accomplish: _20 Minutes_

Like most web apps, our app relies on a combination of remotely fetched data and locally stored data. We can use Apollo Client to manage _both_ types of data, making it a single source of truth for our application's state. We can even interact with both types of data in a single operation. Let's learn how!

<Disclaimer />

## Define a client-side schema

First, let's define a **client-side GraphQL schema** that's specific to our application client. This isn't _required_ for managing local state, but it enables useful developer tooling and helps us reason about our data.

Add the following definition to `src/index.tsx`, before the initialization of `ApolloClient`:

<MultiCodeBlock>

```tsx:title=src/index.tsx
export const typeDefs = gql`
  extend type Query {
    isLoggedIn: Boolean!
    cartItems: [ID!]!
  }
`;
```

</MultiCodeBlock>

Also add `gql` to the list of symbols imported from `@apollo/client`:

```js:title=index.tsx
import {
  ApolloClient,
  NormalizedCacheObject,
  ApolloProvider,

>   gql // highlight-line

Author: We don't, largely because I'm not 100% clear yet on where they're desirable and where they break everything 😄 

But I'll add this one!

} from '@apollo/client';
```

As you might expect, this looks a lot like a definition from our _server's_ schema, with one difference: we **extend** the `Query` type. You can extend a GraphQL type that's defined in another location to add fields to that type.

In this case, we're adding two fields to `Query`:

* `isLoggedIn`, to track whether the user has an active session
* `cartItems`, to track which launches the user has added to their cart

Finally, let's modify the constructor of `ApolloClient` to provide our client-side schema:

<MultiCodeBlock>

```ts:title=src/index.tsx

> const client: ApolloClient<NormalizedCacheObject> = new ApolloClient({

Author: I'm a bit clueless on that point; this is a TypeScript sample that also auto-generates the vanilla JS sample, and my TS chops aren't sufficient to say when this typing is a good and necessary practice versus when it's superfluous 😬

### Enable logout

A logged-in user needs to be able to log out of our client as well. Our example app can perform a logout entirely locally, because logged-in status is determined by the presence of a `token` key in `localStorage`.

Open `src/containers/logout-button.tsx`. Replace its contents with the following:

<ExpansionPanel title="Click to expand">

<MultiCodeBlock>

```tsx{14-25}:title=src/containers/logout-button.tsx
import React from 'react';
import styled from 'react-emotion';
import { useApolloClient } from '@apollo/client';

import { menuItemClassName } from '../components/menu-item';
import { isLoggedInVar } from '../cache';
import { ReactComponent as ExitIcon } from '../assets/icons/exit.svg';

const LogoutButton = () => {
  const client = useApolloClient();
  return (
    <StyledButton
      data-testid="logout-button"
      onClick={() => {
        // Evict and garbage-collect the cached user object
        client.cache.evict({ fieldName: 'me' });
        client.cache.gc();

        // Clear localStorage

>         localStorage.clear();

Author: Agreed, will make corresponding change in "final" version of repo

interface ActionButtonProps extends Partial<LaunchDetailTypes.LaunchDetails_launch> {}

const CancelTripButton: React.FC<ActionButtonProps> = ({ id }) => {
  const [mutate, { loading, error }] = useMutation(
    CANCEL_TRIP,
    {
      variables: { launchId: id },
      update(cache, { data: { cancelTrip } }) {
        // Update the user's cached list of trips to remove the trip that
        // was just canceled.
        const launch = cancelTrip.launches[0];
        cache.modify({

>           id: `User:${localStorage.getItem('userId')}`,

Author: Ahh I'm realizing that I need to make some changes to previous chapters to get `userId` stored locally again. That'll take some time tomorrow.

        onClick={() => mutate()}
        data-testid={'action-button'}
      >
        Cancel This Trip
      </Button>
    </div>
  );
};

> const ToggleTripButton: React.FC<ActionButtonProps> = ({ id }) => {
>   const cartItems = cartItemsVar();
>   const isInCart = id ? cartItems.includes(id) : false;
>   const [, toggleTrip] = useState(isInCart);
>   return (
>     <div>
>       <Button
>         onClick={() => {
>           if (id) {
>             cartItemsVar(
>               isInCart
>                 ? cartItems.filter((i) => i !== id)
>                 : [...cartItems, id]
>             );
>             // Trigger a re-render to pick up the `cartItemsVar` changes.
>             toggleTrip(!isInCart);
>           }
>         }}
>         data-testid={'action-button'}
>       >
>         {isInCart ? 'Remove from Cart' : 'Add to Cart'}

Reviewer: ```suggestion
const ToggleTripButton: React.FC<ActionButtonProps> = ({ id }) => {
  const cartItems = useReactiveVar(cartItemsVar);
  const isInCart = id ? cartItems.includes(id) : false;
  return (
    <div>
      <Button
        onClick={() => {
          if (id) {
            cartItemsVar(
              isInCart
                ? cartItems.filter(itemId => itemId !== id)
                : [...cartItems, id]
            );
          }
        }}
        data-testid={'action-button'}
      >
        {isInCart ? 'Remove from Cart' : 'Add to Cart'}
```
This can be a bit simpler (no need for `useState`) if you use [`useReactiveVar`](https://github.com/apollographql/apollo-client/pull/6867).

As a side note, it might be a good idea to reevaluate `cartItemsVar()` inside the `onClick` handler, so you aren't modifying a stale value. However, in this example, I don't think that causes any problems in practice.
